bin/grep.py

```diff
@@ -27,32 +83,87 @@ def main(args):
 
     pattern = re.compile(ns.pattern, flags=flags)
 
-    # Do not try to grep directories
-    files = [f for f in ns.files if not os.path.isdir(f)]
+    files = []
+    if ns.recursive:
+        for f in ns.files:
+            if f=="-":
+                files.append(f)
+            else:
+                p = Path(f)
+                if p.is_dir():
+                    files += [ str(ff.resolve()) for ff in p.rglob('*') if ff.is_file()]
+                else:
+                    d = p.parent
+                    n = p.name
+                    files += [ str(ff.resolve()) for ff in d.rglob(n) if ff.is_file()]
+    elif len(ns.files)>0:
+        # Do not try to grep directories
+        for f in ns.files:
+            if f=="-":
+                files.append(f)
+            else:
+                p = Path(f)
+                if p.is_file():
+                    files.append(str(p.resolve()))
+                elif not p.exists():
+                    _stash.write_message('{} was skipped because it does not exist'.format(f),prefix="grep: ",error=True)
```

Comment:
I know we use `_stash.write_message()` in some commands, but a `print()` would be better here. `_stash.write_message()` writes directly to the UI while `print()` writes to the threads stdio. Messages outputted using `_stash.write_message()` would not be affected by I/O redirection, e.g. `cat foo.txt | grep bar &> baz.txt` wouldn't work correctly.

Reply:
I only used write_message for error messages from inside the code because those turn up red in the shell window. But I guess I can create some red message via print as well. Is there some way of finding out, if stderr is actually going to the shell window or to some pipe? In the later case I would like to avoid decorating the text with color when printing it.

Reply:
I think you could check via `isinstance(threading.currentThread().sys_stderr, stash.system.shio.ShIO)`, although that would be rather ugly. A somewhat less correct method, but avoiding a `stash.shio` import, would be to check if the class name of `threading.currentThread().sys_stderr` matches ShIO`.

Reply:
Would it make sense to create a printError(txt) method in libcore.py to do it there the right way and then just use that instead of the print(...)? I could do the coding and create a new pull request for that, since it is unrelated to these changes.